### PR TITLE
feat: add mls info to device details view WPB-5520

### PR DESCRIPTION
**Pull request by theswiftway**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-5520" title="WPB-5520" target="_blank">Sub-taskWPB-5520</a>  [iOS] 5.3.23 Show MLS info on devices - UI
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
Adds MLS info on Device Details Screen.
### Issues

When MLS info is available, MLS info is displayed on the Device Details screen.

### Causes (Optional)

_Briefly describe the causes behind the issues. This could be helpful to understand the adopted solutions behind some nasty bugs or complex issues._

### Solutions

Added MLS section to the Device Details screen

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

**Steps**
* Shake the test device to enter Developer Menu
* Tap on Configure flags
* Turn on `enableMLSSupport`
* Close the screen
* Tap on Profile picture
* Tap on Settings
* Tap on Devices
* Tap on any device

Expected: MLS Info should be displayed

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

simulator_screenshot_80139BDE-88D2-4FA6-AD09-9FE86AF6F4EE

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-mono.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 1]
```diff
@@ -1,25 +0,0 @@
...
-        "state": {
-          "branch": null,
-          "revision": "dc46eeb3928a75390651fac6c1ef7f93ad59a73b",
-          "version": "1.11.1"
-        }
-      }
-    ]
-  },
-  "version": 1
-}
```
Comment (netbe) [inline wire-ios-mono.xcworkspace/xcshareddata/swiftpm/Package.resolved]: should not be removed

File: wire-ios/Wire-iOS Tests/DeviceManagement/DeviceDetailsViewTests.swift [hunk 2]
```diff
@@ -80,21 +80,24 @@ final class DeviceDetailsViewTests: ZMSnapshotTestCase, CoreDataFixtureTestHelpe
...
-            isE2EIdentityEnabled: isE2EIdentityEnabled,
-            isSelfClient: isSelfClient,
             userSession: mockSession,
             getUserClientFingerprint: mockGetUserClientFingerprintUseCaseProtocol,
-            userClient: client
+            userClient: client,
+            isSelfClient: isSelfClient
         )
         viewModel.proteusKeyFingerprint = kFingerPrint
+        viewModel.mlsThumbprint = kFingerPrint
```
Comment (netbe) [inline wire-ios/Wire-iOS Tests/DeviceManagement/DeviceDetailsViewTests.swift:92]: what does k prefix stands for `kFingerPrint` ?

File: wire-ios/Wire-iOS/Sources/Developer/DeveloperTools/E2eIdentity/DeveloperDeviceDetailsSettingsSelectionViewModel.swift [hunk 3]
```diff
@@ -101,4 +102,33 @@ class DeveloperDeviceDetailsSettingsSelectionViewModel: ObservableObject {
             Self.selectedE2eIdentiyStatus = item.value
         }
     }
+
+    static func e2eIdentityProvider() -> E2eIdentityProviding {
+       if DeveloperDeviceDetailsSettingsSelectionViewModel.isE2eIdentityViewEnabled {
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/Developer/DeveloperTools/E2eIdentity/DeveloperDeviceDetailsSettingsSelectionViewModel.swift]: what about starting by the return statment with a guard let in order to improve readability
```suggestion
       guard DeveloperDeviceDetailsSettingsSelectionViewModel.isE2eIdentityViewEnabled  else {
```

File: wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/Views/Components/DeviceMLSView.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import SwiftUI
+
+struct DeviceMLSView: View {
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/Views/Components/DeviceMLSView.swift:21]: nice 🫧

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift [hunk 5]
```diff
@@ -97,12 +97,14 @@ final class ClientListViewController: UIViewController,
...
 
-    required init(clientsList: [UserClient]?,
-                  selfClient: UserClient? = ZMUserSession.shared()?.selfUserClient,
-                  credentials: ZMEmailCredentials? = .none,
-                  detailedView: Bool = false,
-                  showTemporary: Bool = true,
-                  showLegalHold: Bool = true) {
+    required init(
+        clientsList: [UserClient]?,
+        selfClient: UserClient? = ZMUserSession.shared()?.selfUserClient,
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift:102]: can we make this non optional? if we don't have a selfUserClient we should not present this screen, no?

**Review by netbe — CHANGES_REQUESTED** [review]
one SPM package should not be removed, otherwise looks good

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift (hunk 5 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift:102]: Not in the scope of this PR, But I will make a note

File: wire-ios/Wire-iOS Tests/DeviceManagement/DeviceDetailsViewTests.swift (hunk 2 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/DeviceManagement/DeviceDetailsViewTests.swift:92]: It's just a common convention to indicate that these are constants .

File: wire-ios/Wire-iOS/Sources/Developer/DeveloperTools/E2eIdentity/DeveloperDeviceDetailsSettingsSelectionViewModel.swift (hunk 3 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/Developer/DeveloperTools/E2eIdentity/DeveloperDeviceDetailsSettingsSelectionViewModel.swift]: done

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift (hunk 5 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift:102]: [WPB-5734](https://wearezeta.atlassian.net/browse/WPB-5734) Made a note here

File: wire-ios-data-model/Source/Model/UserClient/MLSProvider.swift [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public enum MLSError: Error {
+    case decodingError
```
Comment (KaterinaWire) [inline wire-ios-data-model/Source/Model/UserClient/MLSProvider.swift]: I don't see the we use `decodingError` now

File: wire-ios/Wire-iOS Tests/DeviceManagement/DeviceDetailsViewTests.swift [hunk 7]
```diff
@@ -202,55 +221,90 @@ final class DeviceDetailsViewTests: ZMSnapshotTestCase, CoreDataFixtureTestHelpe
...
 
     func testWhenE2eidentityViewIsEnabledAndCertificateIsNotActivatedInDarkMode() {
         let e2eIdentityProvider = MockNotActivatedE2eIdentityProvider()
+        let mlsProvider = MockMLSProvider(isMLSEnbaled: false)
+        verify(
+            matching: setupWrappedInNavigationController(
+                mode: .dark,
+                e2eIdentityProvider: e2eIdentityProvider,
+                mlsProvider: mlsProvider,
+                e2eIdentityCertificate: e2eIdentityProvider.certificate
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS Tests/DeviceManagement/DeviceDetailsViewTests.swift:283]: Why are you passing `e2eIdentityProvider` and `e2eIdentityCertificate` separately if the certificate is part of the `e2eIdentityProvider`?

File: wire-ios/Wire-iOS Tests/DeviceManagement/MockE2eIdentityProvider.swift [hunk 8]
```diff
@@ -27,7 +27,7 @@ public final class MockE2eIdentityProvider: E2eIdentityProviding {
         E2eIdentityCertificate(
             certificateDetails: String(repeating: "abcdefghijklmno", count: 10),
             expiryDate: dateFormatter.date(from: "15.10.2023") ?? Date.now,
-            certificateStatus: "Valid",
+            certificateStatus: "",
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS Tests/DeviceManagement/MockE2eIdentityProvider.swift:30]: `certificateStatus` can't be an empty string

Comment (KaterinaWire) [reply wire-ios/Wire-iOS Tests/DeviceManagement/MockE2eIdentityProvider.swift:30]: In the previous PR I suggested not using it as a string and specifying a specific type. We should prevent this empty state for production code.

Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/DeviceManagement/MockE2eIdentityProvider.swift:30]: I will do the datamodel changes in the business logic PR.

File: wire-ios/Wire-iOS Tests/DeviceManagement/DeviceDetailsViewTests.swift (hunk 7 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/DeviceManagement/DeviceDetailsViewTests.swift:283]: This is just for testing. Snapshot testing doesn't support asynchronous fetching properly.

File: wire-ios/Wire-iOS Tests/DeviceManagement/MockE2eIdentityProvider.swift (hunk 8 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/DeviceManagement/MockE2eIdentityProvider.swift:30]: Noted here https://wearezeta.atlassian.net/browse/WPB-5523

File: wire-ios-data-model/Source/Model/UserClient/MLSProvider.swift (hunk 6 above)
Comment (theswiftway) [reply wire-ios-data-model/Source/Model/UserClient/MLSProvider.swift]: Removed it

File: wire-ios-data-model/Source/Model/UserClient/MLSProvider.swift [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public enum MLSError: Error {
```
Comment (KaterinaWire) [inline wire-ios-data-model/Source/Model/UserClient/MLSProvider.swift:21]: This is a very generic name. As you know, we have a lot of MLS errors and making them public can be confusing.

**Review by netbe — APPROVED** [review]

File: wire-ios-data-model/Source/Model/UserClient/MLSProvider.swift (hunk 9 above)
Comment (theswiftway) [reply wire-ios-data-model/Source/Model/UserClient/MLSProvider.swift:21]: We could change this while implementing the business logic. For now this is just something which is in place

Comment (theswiftway) [reply wire-ios-data-model/Source/Model/UserClient/MLSProvider.swift:21]: Noted here https://wearezeta.atlassian.net/browse/WPB-5521

**Review by KaterinaWire — APPROVED** [review]

**Pull request merged**